File: frontend/packages/contracts/src/wire.ts
import { z } from "zod";

/** Wire = News/OSINT & gov-feed terminal. Each desk is a column. */
export const WIRE_DESKS = [
  "intel",
  "world",
  "middle-east",
  "europe",
  "africa",
  "latin-america",
  "asia-pacific",
  "united-states",
  "markets",
  "corporate",
] as const;
export type WireDesk = (typeof WIRE_DESKS)[number];

export const WIRE_DESK_LABELS: Record<WireDesk, string> = {
  intel: "Intel Feed",
  world: "World News",
  "middle-east": "Middle East",
  europe: "Europe",
  africa: "Africa",
  "latin-america": "Latin America",
  "asia-pacific": "Asia-Pacific",
  "united-states": "United States",
  markets: "Markets & Macro",
  corporate: "Corporate Wire",
};

export const WIRE_CATEGORIES = [
  "conflict",
  "disaster",
  "protest",
  "diplomatic",
  "economic",
  "general",
] as const;
export type WireCategory = (typeof WIRE_CATEGORIES)[number];

export const WIRE_SEVERITIES = ["breaking", "alert", "caution", "normal"] as const;
export type WireSeverity = (typeof WIRE_SEVERITIES)[number];

export const WireItem = z
  .object({
    title: z.string(),
    link: z.string(),
    /** ISO 8601 when parseable, else raw, else "". */
    pubDate: z.string(),
    summary: z.string(),
    source: z.string(),
    sourceId: z.string(),
    desk: z.enum(WIRE_DESKS),
    // Heuristic tags — kept as loose strings so a server-side taxonomy tweak
    // never fails the client parse.
    category: z.string(),
    severity: z.string(),
    // Exchange-tagged symbols named in the headline (Corporate Wire desk).
    // nullish so older cached payloads / other desks parse cleanly.
    tickers: z.array(z.string()).nullish(),
  })
  .passthrough();
export type WireItem = z.infer<typeof WireItem>;

export const WireItemsResponse = z.object({
  desk: z.enum(WIRE_DESKS),
  items: z.array(WireItem),
  lastUpdated: z.string().nullish(),
});
export type WireItemsResponse = z.infer<typeof WireItemsResponse>;

export const WireBreakingResponse = z.object({
  items: z.array(WireItem),
  lastUpdated: z.string().nullish(),
});
export type WireBreakingResponse = z.infer<typeof WireBreakingResponse>;

export const WireDesksResponse = z.object({
  desks: z.array(
    z.object({
      id: z.enum(WIRE_DESKS),
      label: z.string(),
      sources: z.array(z.string()),
    }),
  ),
  totalFeeds: z.number().nullish(),
  lastUpdated: z.string().nullish(),
});
export type WireDesksResponse = z.infer<typeof WireDesksResponse>;

File: moby/lib/data/models/wire.dart
// Wire = News/OSINT & gov-feed terminal. Mirrors the web contract
// (`frontend/packages/contracts/src/wire.ts`) 1:1 — keep desks, labels, and
// the item shape in sync with it (data-display parity is a hard rule).

/// The ten Wire desks, in the canonical order the contract declares them.
/// `id` is the server's desk slug (used as the `?desk=` query param); `label`
/// is the display name shown on the desk chips.
enum WireDesk {
  intel,
  world,
  middleEast,
  europe,
  africa,
  latinAmerica,
  asiaPacific,
  unitedStates,
  markets,
  corporate;

  String get id => switch (this) {
        WireDesk.intel => 'intel',
        WireDesk.world => 'world',
        WireDesk.middleEast => 'middle-east',
        WireDesk.europe => 'europe',
        WireDesk.africa => 'africa',
        WireDesk.latinAmerica => 'latin-america',
        WireDesk.asiaPacific => 'asia-pacific',
        WireDesk.unitedStates => 'united-states',
        WireDesk.markets => 'markets',
        WireDesk.corporate => 'corporate',
      };

  String get label => switch (this) {
        WireDesk.intel => 'Intel Feed',
        WireDesk.world => 'World News',
        WireDesk.middleEast => 'Middle East',
        WireDesk.europe => 'Europe',
        WireDesk.africa => 'Africa',
        WireDesk.latinAmerica => 'Latin America',
        WireDesk.asiaPacific => 'Asia-Pacific',
        WireDesk.unitedStates => 'United States',
        WireDesk.markets => 'Markets & Macro',
        WireDesk.corporate => 'Corporate Wire',
      };

  static WireDesk? fromId(String id) {
    for (final d in WireDesk.values) {
      if (d.id == id) return d;
    }
    return null;
  }
}

/// A single classified headline. `category` / `severity` are kept as loose
/// strings (the server taxonomy can change without breaking the client parse),
/// mirroring the contract's `.passthrough()`.
class WireItem {
  const WireItem({
    required this.title,
    required this.link,
    required this.pubDate,
    required this.summary,
    required this.source,
    required this.sourceId,
    required this.desk,
    required this.category,
    required this.severity,
    required this.tickers,
  });

  final String title;
  final String link;

  /// ISO 8601 when parseable, else raw, else "".
  final String pubDate;
  final String summary;
  final String source;
  final String sourceId;
  final String desk;
  final String category;
  final String severity;

  /// Exchange-tagged symbols named in the headline (Corporate Wire desk in
  /// practice). Empty for other desks.
  final List<String> tickers;

  factory WireItem.fromJson(Map<String, dynamic> j) => WireItem(
        title: j['title'] as String? ?? '',
        link: j['link'] as String? ?? '',
        pubDate: j['pubDate'] as String? ?? '',
        summary: j['summary'] as String? ?? '',
        source: j['source'] as String? ?? '',
        sourceId: j['sourceId'] as String? ?? '',
        desk: j['desk'] as String? ?? '',
        category: (j['category'] as String?)?.isNotEmpty == true
            ? j['category'] as String
            : 'general',
        severity: (j['severity'] as String?)?.isNotEmpty == true
            ? j['severity'] as String
            : 'normal',
        tickers: (j['tickers'] as List?)
                ?.map((e) => e.toString())
                .where((s) => s.isNotEmpty)
                .toList() ??
            const <String>[],
      );

  /// Milliseconds since epoch, or null when the date is unparseable.
  int? get publishedAtMs => DateTime.tryParse(pubDate)?.millisecondsSinceEpoch;
}

/// Items for one desk plus the server's freshness stamp.
class WireDeskItems {
  const WireDeskItems({required this.items, required this.lastUpdated});
  final List<WireItem> items;
  final String? lastUpdated;
}
